### PR TITLE
 enhance(apps/frontend-pwa): display student response scoring in learning element progress bar

### PR DESCRIPTION
**Pull request by sjschlapbach**
This pull request adds status highlighting to the learning element progress bar depending on the students answers:
- If all elements in a stack are answered correctly, the step will be highlighted as correct (green)
- If at least one element in a stack was only partially correct, the step will be highlighted as partially correct (orange)
- If all elements of a stack were answered incorrectly, the step will be highlighted as incorrect (red)
- Stacks with text elements only will be automatically highlighted as correct as soon as the user has navigated there

Screenshot 2023-08-12 at 17 43 50

### CODE REVIEW

File: apps/frontend-pwa/src/components/learningElements/LearningElement.tsx [hunk 1]
```diff
@@ -19,6 +20,9 @@ function LearningElement({
   handleNextQuestion,
 }: LearningElementProps) {
   const currentStack = element.stacks?.[currentIx]
+  const [stepStatus, setStepStatus] = useState<
+    ('unanswered' | 'incorrect' | 'partial' | 'correct')[]
+  >(Array.from({ length: element.stacks?.length ?? 0 }, () => 'unanswered'))
```
Comment (rschlaefli) [inline apps/frontend-pwa/src/components/learningElements/LearningElement.tsx]: might be easier to read to use functions from ramda for such cases, e.g.,

```suggestion
  R.repeat('unanswered', element.stacks?.length ?? 0)
```

File: apps/frontend-pwa/src/components/learningElements/QuestionStack.tsx [hunk 2]
```diff
@@ -143,16 +146,58 @@ function QuestionStack({
...
+          !element.questionInstance || element.questionInstance.evaluation
+      ) ??
+      false
+    ) {
+      setIsEvaluation(true)
+    }
+
+    let overallScore: 'correct' | 'incorrect' | 'partial' | 'unanswered' =
+      'unanswered'
+    stack.elements?.forEach((element) => {
```
Comment (rschlaefli) [inline apps/frontend-pwa/src/components/learningElements/QuestionStack.tsx]: maybe generalize this logic and use reduce

File: packages/graphql/src/services/learningElements.ts [hunk 3]
```diff
@@ -96,6 +97,7 @@ function evaluateQuestionResponse(
             pointsPercentage,
             multiplier: multiplier ?? 1,
           }),
+          percentile: pointsPercentage ?? 0,
```
Comment (rschlaefli) [inline packages/graphql/src/services/learningElements.ts:100]: should we add these defaults in the grading package or does it make more sense here?

Comment (sjschlapbach) [reply packages/graphql/src/services/learningElements.ts:100]: After checking the implementation again, I think it makes sense to leave it like this as 0 and null have different meanings in the grading package. Null is only returned, if no sample solutions are specified (i.e. content cannot be graded) and 0 is only returned when the answer was entirely wrong. Even though the case that there are no sample solutions cannot occur for learning elements (as we do not allow the use of such questions), it probably makes sense to still distinguish these cases in the grading package.

File: apps/frontend-pwa/src/components/learningElements/LearningElement.tsx (hunk 1 above)
Comment (sjschlapbach) [reply apps/frontend-pwa/src/components/learningElements/LearningElement.tsx]: good idea, changed it

File: apps/frontend-pwa/src/components/learningElements/QuestionStack.tsx (hunk 2 above)
Comment (sjschlapbach) [reply apps/frontend-pwa/src/components/learningElements/QuestionStack.tsx]: the entire logic has been reworked into a single and more concise useEffect snipped with reduced complexity and better readability

**Review by rschlaefli — APPROVED** [review]

**Comment by rschlaefli** [comment]
Extended version including points/scoring and icon for content-only parts (using custom formatter on StepProgress component)

MicrosoftTeams-image (1)

**Pull request merged**